src/java/org/apache/cassandra/utils/btree/BTree.java

```diff
@@ -219,103 +214,92 @@ public static Dir desc(boolean desc)
+            int threshold = denseChildSize + 1 + MIN_KEYS * (denseGrandChildSize + 1);
+            int remaining = size;
             int i = 0;
-            while (i < denseCount)
-            {
-                sizeMap[i] = (1 + i) * (1 + denseSize(height)) - 1;
-                branch[keyCount + i] = buildPerfectDense(source, height, updateF);
-                branch[i++] = isSimple(updateF) ? source.next() : updateF.apply(source.next());
-            }
-            int sparseCount = childCount - (1 + denseCount);
-            int remainderSize = size - (denseCount * (1+denseSize(height)) + sparseCount * (1+sparseSize(height)));
+            // Add dense children until we reach the threshold.
+            while (remaining >= threshold)
             {
-                // need to decide our number of children either based on denseGrandChildSize,
-                // or the least number of children we can use, if we want to be maximally dense
-                int grandChildCount = remainderSize < denseSize(height)/2
-                                      ? MIN_KEYS + 1
-                                      : remainderSize / (denseSize(height-1)+1) + 1;
-
-                branch[keyCount + i] = buildMaximallyDense(source, grandChildCount, remainderSize, height, updateF);
-                sizeMap[i] = size = (1 + denseSize(height)) * i + remainderSize;
+                branch[keyCount + i] = buildPerfectDenseAndAddUsage(source, height, updateF);
+                branch[i] = isSimple(updateF) ? source.next() : updateF.apply(source.next());
+                remaining -= denseChildSize + 1;
+                sizeMap[i++] = size - remaining - 1;
             }
-            // finally build sparse children to consume the remaining elements
-            while (i < keyCount)
+            // If the remainder does not fit one child, split it roughly in half, where the first child only has dense
+            // grandchildren.
+            if (remaining > denseChildSize)
             {
-                branch[i++] = isSimple(updateF) ? source.next() : updateF.apply(source.next());
-                branch[keyCount + i] = buildPerfectSparse(source, height, updateF);
-                sizeMap[i] = size += 1 + sparseSize(height);
-            }
-            if (!isSimple(updateF))
-            {
-                // accounts only for the perfect allocations; recursive calls of buildMaximallyDense will account for the remainder
-                updateF.onAllocated(denseCount * PERFECT_DENSE_SIZE_ON_HEAP[height - 1]
-                                    + sparseCount * PERFECT_SPARSE_SIZE_ON_HEAP[height - 1]);
+                int grandChildCount = remaining / ((denseGrandChildSize + 1) * 2);
+                assert grandChildCount >= MIN_KEYS + 1;
+                int childSize = grandChildCount * (denseGrandChildSize + 1) - 1;
+                branch[keyCount + i] = buildMaximallyDense(source, grandChildCount, childSize, height, updateF);
```

Comment:
could we use `buildPerfectDenseAndAddUsage` here?

Reply:
I think I made a mistake in my thinking anyway, and this was fine. I'll have a look at the new code once it's pushed.

Reply:
Ok, I found the version in your repository. I think roll this back, it was my mistake when suggesting it and this isn't worth any clarity penalty.